File: goLambda/internal/usecases/usecases.go
package usecases

import (
	"fmt"
	"math/rand"
	"sync"

	"github.com/Abraxas-365/lambda-example/pkg/pokemon"
	"github.com/Abraxas-365/lambda-example/pkg/user"
)

type Usecases struct {
	userRepo    user.UserRepository
	pokemonRepo pokemon.PokemonRepository
}

func New(userRepo user.UserRepository, pokemonRepo pokemon.PokemonRepository) *Usecases {
	return &Usecases{
		userRepo:    userRepo,
		pokemonRepo: pokemonRepo,
	}
}

func (u *Usecases) CreateUser(user user.User) error {
	return u.userRepo.Save(user)
}

func (u *Usecases) GetUserPokemons(userId string) ([]pokemon.Pokemon, error) {
	pokemonList := []string{
		"bulbasaur", "ivysaur", "venusaur", "charmander",
		"charmeleon", "charizard", "squirtle", "wartortle",
		"blastoise", "caterpie", "metapod", "butterfree",
		"weedle", "kakuna", "beedrill", "pidgey", "pidgeotto",
		"pidgeot", "rattata", "raticate",
	}

	user, err := u.userRepo.FindById(userId)
	if err != nil {
		return nil, err
	}
	fmt.Println(user)

	var wg sync.WaitGroup

	resultCh := make(chan pokemon.Pokemon, user.NPokeBall)
	errorCh := make(chan error, user.NPokeBall)

	for i := 0; i < user.NPokeBall; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()

			randomIndex := rand.Intn(len(pokemonList))

			pokemonName := pokemonList[randomIndex]
			pokemon, err := u.pokemonRepo.FindByName(pokemonName)
			if err != nil {
				errorCh <- err
				return
			}
			resultCh <- pokemon
		}()
	}

	go func() {
		wg.Wait()
		close(resultCh)
		close(errorCh)
	}()

	select {
	case err := <-errorCh:
		return nil, err
	default:
	}

	var pokemons []pokemon.Pokemon
	for i := 0; i < user.NPokeBall; i++ {
		pokemon := <-resultCh
		pokemons = append(pokemons, pokemon)
	}

	return pokemons, nil
}

File: rust_lambda/backend/src/use_cases.rs
use futures;
use lib::{
    errors::errors::Error,
    pokemon::{pokemon_repository::PokemonRepository, pokemon_type::Pokemon},
    user::{user_repository::UserRepository, user_type::User},
};
use rand::Rng;
use tokio;

pub struct UseCases {
    user_repo: Box<dyn UserRepository>,
    pokemos_repo: Box<dyn PokemonRepository>,
}

impl UseCases {
    pub fn new(
        user_repo: Box<dyn UserRepository>,
        pokemos_repo: Box<dyn PokemonRepository>,
    ) -> Self {
        Self {
            user_repo,
            pokemos_repo,
        }
    }

    pub async fn create_user(&self, user: User) -> Result<(), Error> {
        self.user_repo.save(user).await?;
        Ok(())
    }
    pub async fn get_user_pokemons(&self, user_id: &str) -> Result<Vec<Pokemon>, Error> {
        let pokemons_name = vec![
            "bulbasaur".to_string(),
            "ivysaur".to_string(),
            "venusaur".to_string(),
            "charmander".to_string(),
            "charmeleon".to_string(),
            "charizard".to_string(),
            "squirtle".to_string(),
            "wartortle".to_string(),
            "blastoise".to_string(),
            "caterpie".to_string(),
            "metapod".to_string(),
            "butterfree".to_string(),
            "weedle".to_string(),
            "kakuna".to_string(),
            "beedrill".to_string(),
            "pidgey".to_string(),
            "pidgeotto".to_string(),
            "pidgeot".to_string(),
            "rattata".to_string(),
            "raticate".to_string(),
        ];
        let user = self.user_repo.find_by_id(&user_id).await?;

        let mut user_pokemon_futures: Vec<_> = Vec::new();

        for _ in 0..user.number_of_pokeballs {
            let random_index = rand::thread_rng().gen_range(0..pokemons_name.len());
            let pokemon_future = self
                .pokemos_repo
                .find_by_name(pokemons_name[random_index].as_str());
            user_pokemon_futures.push(pokemon_future);
        }

        let user_pokemons: Result<Vec<_>, _> =
            futures::future::try_join_all(user_pokemon_futures).await;

        match user_pokemons {
            Ok(pokemons) => Ok(pokemons),
            Err(err) => Err(err),
        }
    }
}
